🌱 Allow Control Plane endpoint deferral assignment

prometherion opened the pull request:
**What this PR does / why we need it**:

This PR drafts the support for the provisioning of Hetzner-based worker nodes using an externally managed Control Plane.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #845

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

guettli commented on api/v1beta1/hetznercluster_webhook.go:
```diff
@@ -20,12 +20,13 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	"github.com/syself/cluster-api-provider-hetzner/pkg/utils"
 )
```
@prometherion just out of curiosity, why did you re-order the imports? Which tool do you use for this? If it makes sense, we could integrate this in our IDE/CI, too.

prometherion replied:
I'm using `goimports` locally with the `-l` flag.

It seems already enabled in the `golangci-lint` file, I'll configure it properly and open a different PR to keep this atomic to the issue mentioned.

prometherion replied:
#849

guettli commented on controllers/hetznercluster_controller.go:
```diff
@@ -194,7 +193,9 @@ func (r *HetznerClusterReconciler) reconcileNormal(ctx context.Context, clusterS
 	conditions.MarkTrue(hetznerCluster, infrav1.PlacementGroupsSynced)
 
-	if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled {
+	if !hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled && (hetznerCluster.Spec.ControlPlaneEndpoint == nil || len(hetznerCluster.Spec.ControlPlaneEndpoint.Host) == 0) {
```
nit: `len(hetznerCluster.Spec.ControlPlaneEndpoint.Host) == 0` since Host is a string, I would prefer `... == ""`. Or do you have special reasons to check for the length?

prometherion replied:
I'm just used to checking the length of strings, I'll cope with your coding style!

guettli reviewed (commented):
Were you able to integrate caph with these changes into Kamaji? Or is this just the start and further PRs are needed?

prometherion commented:
I was, @guettli

```
NAME                                                        READY  SEVERITY  REASON  SINCE  MESSAGE                                                                        
Cluster/workload                                            True                     11m                                                                                    
├─ClusterInfrastructure - HetznerCluster/workload                                                                                                                           
├─ControlPlane - KamajiControlPlane/workload-control-plane                                                                                                                  
└─Workers                                                                                                                                                                   
  └─MachineDeployment/workload-md-0                         True                     3m8s                                                                                   
    └─2 Machines...                                         True                     10m    See workload-md-0-5849b98d48xhd4bc-lrlhc, workload-md-0-5849b98d48xhd4bc-txq4j
```

With the changes introduced with 7f7ace86c26d806808a33394e3b3e8b2d63589e3, I was able to create the externally managed Control Plane, with the worker nodes on Hetzner Cloud.

I had some issues with the `HetznerCluster.spec.controlPlaneRegions` field which is required:
https://github.com/syself/cluster-api-provider-hetzner/blob/07cb7fb379ba30412bc52a74103faa2a6e02902b/config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerclusters.yaml#L283-L286

With the feature of externally managed Control Planes, that field become obsolete, although the API is still demanding it.

A possible solution is to mark is optional thanks to the struct tag `,omitempty`:

https://github.com/syself/cluster-api-provider-hetzner/blob/07cb7fb379ba30412bc52a74103faa2a6e02902b/api/v1beta1/hetznercluster_types.go#L37-L40

However, once applied the tag, the resulting manifests generated with `make generate-manifests` where a bit outdated:

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   api/v1beta1/hetznercluster_types.go
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hcloudmachines.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hcloudmachinetemplates.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hcloudremediations.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hcloudremediationtemplates.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerbaremetalhosts.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerbaremetalmachines.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerbaremetalmachinetemplates.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerbaremetalremediations.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerbaremetalremediationtemplates.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerclusters.yaml
        modified:   config/crd/bases/infrastructure.cluster.x-k8s.io_hetznerclustertemplates.yaml
        modified:   config/rbac/role.yaml
        modified:   config/webhook/manifests.yaml
```

Let me know what's the best strategy to overcome this.

prometherion commented:
Thanks to #849 now the changes are referring only to the code base, imports are now automatically sorted using `gci`.

May I ask for a review, or a discussion on this, please?

guettli commented:
@prometherion we are working on a product launch at the moment. Sorry, it will take some days until we have time to check the PR.

katlol commented:
@batistein @guettli would you know what the status on this is?

prometherion commented:
Dunno if the PR has been marked as __draft__ due to the conflicts, which are now solved since I rebased the PR over the latest `main` branch.

https://github.com/syself/cluster-api-provider-hetzner/pull/848#issuecomment-1632532332 is still unanswered, tho.

If you have a community meeting I'm happy to join it and explain the benefits of this improvement, as well as closing the remarks to get this merged.

prometherion commented:
I'm a bit puzzled by this feedback, besides putting the PR in draft mode without having a proper discussion.
However, your project, your ownership, you decide on it.

Other infrastructure providers are already offering a deferred Control Plane address assignment, such as:

- OpenStack
- AWS (although a WIP)
- Metal3

The concept of an externally managed control plane means the Control Plane endpoint will be provided externally, and this can happen in two ways:

- a manual assignment
- a Control Plane provider will patch the Cluster definition

Besides the technical consideration, Cluster API allows to have externally managed control planes, having an empty `controlPlaneEndpoint` object is accepted by several other providers, by accepting a bare `controlPlaneAddress` object, or by providing an external field such as `OpenStackCluster.spec.apiServerFixedIP`.

Rather than doing tricky things in the operator with args or feature flags, I would suggest having a special annotation in the `HetznerCluster` object such as `capi.syself.com/allow-empty-control-plane-address` (YMMV).

A user, by creating an `HetznerCluster` with that annotation, would be allowed to skip the opinionated enforcement in non-accepting an empty string: we're not breaking the default approach, a user must specify that value, and whoever wants to use an external control plane provider is happy. A win-win scenario to me.

janiskemper commented:
I don't get why it puzzled you so much that this PR has been set to draft mode. I don't think @batistein spent too much thought on it. He probably did it mainly because we create all our PRs in draft mode until they are ready to be merged. 

This one clearly is not, as we are still in the process of discussing how to approach this scenario.

I also don't understand why you believe that we don't want to support this feature, as I just gave an idea of how to implement it. You gave another idea with the annotation and this sounds completely fine to me, as it serves the same purpose but also allows users to both create clusters with and without this feature and keep the strong validation we have now. 

I am okay with proceeding that way!

guettli commented:
@prometherion what are your plans here? Do you want to change the PR or do you think it should be merged like this?

janiskemper commented:
implemented in #1106
